### PR TITLE
Raspberry Pi OIDC account integration

### PR DESCRIPTION
**Pull request by mwtrew**
---

**Pull Request Check List**
- [x] Unit Tests & Regression Tests Added (Optional)
- [x] Removed Unnecessary Logs/System.Outs/Comments/TODOs
- ~~Added enough Logging to monitor expected behaviour change~~
- ~~Security - Injection - everything run by an interpreter (SQL, OS...) is either validated or escaped~~
- ~~Security - Data Exposure - PII is not stored or sent unencrypted~~
- [x] Security - Data Exposure - Test any altered or created endpoints using [swagger](http://localhost:8080/isaac-api/api-docs/)
- ~~Security - Access Control - Check authorisation on every new endpoint~~
- ~~Security - New dependency - configured sensibly not relying on defaults~~
- ~~Security - New dependency - Searched for any know vulnerabilities~~
- ~~Security - New dependency - Signed up team to mailing list~~
- ~~Security - New dependency - Added to dependency list~~
- ~~DB schema changes - postgres-rutherford-create-script updated~~
- ~~DB schema changes - upgrade script created matching create script~~
- ~~Updated Release Procedure & Documentation (& Considered Implications to Previous Versions)~~
- [x] Peer-Reviewed

### CODE REVIEW

**Review by mlt47 — COMMENTED** [review]
Generally looks good and neat. A few small suggestions so far. I'll have another quick look tomorrow but unlikely to bring up anything else and I didn't want to be blocking you.

File: src/main/java/uk/ac/cam/cl/dtg/segue/auth/RaspberryPiOidcAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+     * our given name/family name format. There is no way to consistently get this right across name structures, so the
+     * user should have an opportunity to correct it later.
+     *
+     * @param nickname The nickname from the IdP.
+     * @param tokenisedName The tokenised full name from the IdP.
+     * @return A list with two elements: the derived given name and family name.
+     * @throws NoUserException when the derived names are not valid.
+     */
+    public List<String> getGivenNameFamilyName(String nickname, List<String> tokenisedName) throws NoUserException {
+        String givenName = nickname;
```
[the remark this answers is not in the document]
Comment (mwtrew) [reply src/main/java/uk/ac/cam/cl/dtg/segue/auth/RaspberryPiOidcAuthenticator.java:259]: UI-wise the field is labelled "What should we call you?", based on that I would expect to see calling names more often than pseudonyms. I might be able to find out.

File: src/main/java/uk/ac/cam/cl/dtg/segue/auth/RaspberryPiOidcAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            // If the full name is empty, use the nickname in both fields. Validation on the IdP side should prevent this.
+            familyName = nickname;
+        } else {
+            // Otherwise, use the last token of the full name as the family name.
+            familyName = tokenisedName.get(tokenisedName.size() - 1);
+        }
+
+        // Finally, check that the name meets validation.
+        if (!UserAccountManager.isUserNameValid(givenName) || !UserAccountManager.isUserNameValid(familyName)){
+            throw new NoUserException("The name provided by the identity provider does not meet validation.");
```
[the remark this answers is not in the document]
Comment (mwtrew) [reply src/main/java/uk/ac/cam/cl/dtg/segue/auth/RaspberryPiOidcAuthenticator.java:272]: I think making sure IDP-provided names are subject to the same validation as the default sign-up flow is sensible for the sake of consistency. If this validation isn't doing anything useful in general I'd argue we should re-visit it.

> They will get a chance to correct it on the required account modal soon after sign up (we might still need to create the Trello card for this!).

I've just created a card for this, but I think it's something to come back to when we have a bit more time.

[the remark this answers is not in the document]
Comment (mwtrew) [reply src/main/java/uk/ac/cam/cl/dtg/segue/auth/RaspberryPiOidcAuthenticator.java:278]: > Generally looks good and neat. A few small suggestions so far. I'll have another quick look tomorrow but unlikely to bring up anything else and I didn't want to be blocking you.

In that case I'll go ahead and merge this, do let me know if you have any other thoughts and I'll come back to it.

**Comment by mwtrew** [comment]
> Generally looks good and neat. A few small suggestions so far. I'll have another quick look tomorrow but unlikely to bring up anything else and I didn't want to be blocking you.

**Pull request merged**